packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -245,41 +260,292 @@ class FilteringTextInputFormatter extends TextInputFormatter {
+      if (allow) {
+        // The non-match region is the banned pattern.
+        allowStart = match.start;
+        allowEnd = match.end;
+        replaceStart = previousMatch?.end ?? 0;
+        replaceEnd = match.start;
+
+        // Apply the replacingStrategy first since the non-match region is
+        // banned.
+        replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+        accumulator.stringBuffer.write(newValue.text.substring(allowStart, allowEnd));
+      } else {
+        // When `allow` is false the match region is the banned pattern.
+        allowStart = previousMatch?.end ?? 0;
+        allowEnd = match.start;
+        replaceStart = match.start;
+        replaceEnd = match.end;
+
+        // Write the allowed word to the buffer first.
+        accumulator.stringBuffer.write(newValue.text.substring(allowStart, allowEnd));
+        replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+      }
+
+      previousMatch = match;
+    }
+
+    // Handle the last non-match region betwee the last match region and the end
+    // of the text.
+    final int replaceStart = previousMatch?.end ?? 0;
+    final int replaceEnd = newValue.text.length;
+
+    if (allow) {
+      replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+    } else {
+      accumulator.stringBuffer.write(newValue.text.substring(replaceStart));
+    }
+
+    return accumulator._finalize();
   }
 
   /// A [TextInputFormatter] that forces input to be a single line.
   static final TextInputFormatter singleLineFormatter = FilteringTextInputFormatter.deny('\n');
 
   /// A [TextInputFormatter] that takes in digits `[0-9]` only.
   static final TextInputFormatter digitsOnly = FilteringTextInputFormatter.allow(RegExp(r'[0-9]'));
+
+  /// A [FilteringFormatterReplacingStrategy] that attempts to preserve the
+  /// pre-format selection and composing region as much as possible.
+  ///
+  /// If the start or the end of the selection (or the composing region) is
+  /// strictly inside the banned pattern described by `regionStart` and
+  /// `regionEnd` (meaning, `regionStart` < index < `regionEnd`), the index will
+  /// be placed at the end of the replacement string.
+  static void preserveSelectionAndComposingRegion(
+    TextEditingValueAccumulator accumulator,
+    int regionStart,
+    int regionEnd,
+    String replacement,
+  ) {
+    assert(regionEnd >= regionStart);
+    if (regionEnd == regionStart) {
+      return;
+    }
+
+    final MutableTextRange<TextSelection>? selection = accumulator.selection;
+    final MutableTextRange<TextRange>? composingRegion = accumulator.composingRegion;
+
+    // Always replace the text.
+    accumulator.stringBuffer.write(replacement);
+
+    final int replacementLength = replacement.length;
+    // Don't modify the text ranges if the region is replaced with a string of
+    // the same length.
+    if (regionEnd - regionStart == replacementLength) {
```

Comment:
Not sure we want this. It kinda makes sense if the replacement string is `*`, but If the replacement string is something like `<redacted>`, and the banned pattern happens to have the same length as the replacement then this is a bit strange.

Reply:
Removed the special case.